File: examples/llm_stream.rs
use jeeves_core::prelude::*;
use std::sync::Arc;

#[tokio::main]
async fn main() -> Result<()> {
    let llm = Arc::new(MockLlmProvider::new([vec![
        ModelStreamEvent::Text("Detective Mara ".into()),
        ModelStreamEvent::Text("found the ledger ".into()),
        ModelStreamEvent::Text("behind the clock tower.".into()),
    ]]));

    let workflow = Workflow::builder("briefing")
        .stage(Stage::llm(
            "speak",
            LlmAction::text(Prompt::text(
                "Summarize the case file in one dramatic sentence.",
            )),
        ))
        .build()?;

    let engine = Engine::builder().llm(llm).workflow(workflow)?.build()?;
    let mut handle = engine.start("briefing", "case 41: the missing ledger")?;

    let mut events = handle
        .take_events()
        .ok_or_else(|| Error::internal("event stream already taken"))?;

    let mut final_outcome = None;
    while let Some(event) = events.recv().await {
        match event {
            RunEvent::TextDelta { content, .. } => print!("{content}"),
            RunEvent::Finished(outcome) => final_outcome = Some(outcome),
            _ => {}
        }
    }

    println!();
    let outcome =
        final_outcome.ok_or_else(|| Error::internal("run ended without a Finished event"))?;
    println!("completed: {}", outcome.completed());
    println!("model calls: {}", outcome.result().usage.llm_calls);

    Ok(())
}

File: examples/llm_stream.cpp
#include <jeeves/jeeves.hpp>

#include <iostream>

using namespace jeeves;

Result<void> briefing() {
    auto llm = std::make_shared<MockLlmProvider>(std::vector<std::vector<ModelStreamEvent>>{{
        "Detective Mara ", "found the ledger ", "behind the clock tower."}});
    auto workflow = Workflow::builder("briefing").stage(Stage::llm("speak", LlmAction::text(
        Prompt::text("Summarize the case file in one dramatic sentence.")))).build();
    if (!workflow) return std::unexpected(workflow.error());
    auto registered = Engine::builder().llm(llm).workflow(std::move(*workflow));
    if (!registered) return std::unexpected(registered.error());
    auto engine = registered->build();
    if (!engine) return std::unexpected(engine.error());
    auto handle = engine->start("briefing", "case 41: the missing ledger");
    if (!handle) return std::unexpected(handle.error());
    auto events = handle->take_events();
    while (auto event = events->recv())
        if (const auto * delta = std::get_if<RunEvent::TextDelta>(&event->value)) std::cout << delta->content << std::flush;
    auto outcome = handle->result();
    if (!outcome) return std::unexpected(outcome.error());
    std::cout << "\ncompleted: " << (*outcome)->completed() << "\nmodel calls: " << (*outcome)->result().usage.llm_calls << '\n';
    if (!(*outcome)->completed()) return std::unexpected(Error::permanent("briefing did not complete"));
    return {};
}

int main() {
    auto result = briefing();
    if (!result) { std::cerr << result.error() << '\n'; return 1; }
}
